### PR TITLE
Allow user to merge datastore-indexes-auto.xml changes

### LINKED ISSUE #2309: datastore-indexes-auto.xml ignored in deploy
- Cloud Tools for Eclipse version: 1.3.0.201708031339
- Google Cloud SDK version: 166.0.0
- OS: Mac OS X 10.12.5
- Java version: 1.8.0_121

**What did you do?**
Running the Google Cloud Tools for Eclipse and successfully deploying to App Engine Standard. The code is correctly published to the app engine cloud and generally works, but indexes (that are correctly auto-generated by the dev server) are not copied up to the cloud standard. I can manually add indexes to the datastore-indexes.xml and they are created, but auto-generated indexes are not.

My datastore-indexes.xml looks like so --->>
```
<?xml version="1.0" encoding="utf-8"?>
<datastore-indexes autoGenerate="true">
    <datastore-index kind="QuestionData" ancestor="false" source="manual">
        <property name="pid" direction="asc"/>
        <property name="order" direction="asc"/>
    </datastore-index>
</datastore-indexes>
```

My datastore-indexes-auto.xml file looks like so --->>

```
<!-- Indices written at Sun, 20 Aug 2017 10:15:28 AEST -->
<datastore-indexes>
    <!-- Used 1 time in query history -->
    <datastore-index kind="QuestionData" ancestor="false" source="auto">
        <property name="envscope" direction="asc"/>
        <property name="pid" direction="asc"/>
        <property name="order" direction="asc"/>
    </datastore-index>
</datastore-indexes>
```

**What did you expect to see?**
That the datastore-indexes-auto.xml would be used to create additional indexes in my datastore.

**What did you see instead?**
No indexes are created, and it would seem the temp location, such as
 workspace/.metadata/.plugins/org.eclipse.wst.server.core/tmp0/ that is used by the Dev Server isn't seen or used by the Cloud Tools to Deploy, and so indexes are not created as defined in the datastore-indexes-auto.xml file that is created by the Dev Server running.

### PR DESCRIPTION
**Pull request by briandealwis**
When the user stops the server, we check if the devappserver created an `datastore-indexes-auto.xml` and, if found and more recent, then prompt the user as to whether they would like to review the changes.

screen shot 2018-05-04 at 1 28 38 am

The changes are shown using the Eclipse Compare mechanism.  This is just using the normal text compare: the user can copy changes over, and make changes directly to their `datastore-indexes.xml` too.

screen shot 2018-05-04 at 1 27 40 am

As we don't delete the devappserver-generated `datastore-indexes-auto.xml` and it is not otherwise removed, we compare the timestamps to avoid unnecessary prompting.

Since the generated `-auto` isn't a perfect update, it might be worth a deeper dive into Eclipse Compare to configure a structural comparator to allow copying the individual entities and properties.  But that's substantially more work.

Fixes #2309

### CODE REVIEW

File: plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/messages.properties [hunk 1]
```diff
@@ -10,6 +10,8 @@ CREATE_APP_ENGINE_RUNTIME_WIZARD_DESCRIPTION=The App Engine Standard runtime req
 CREATE_APP_ENGINE_RUNTIME_WIZARD_TITLE=App Engine Standard Runtime
 SERVICE_CONFLICTS=Cannot add "{0}" as project "{1}" has App Engine Service ID: {2}
 OPEN_CLOUD_SDK_PREFERENCE_BUTTON=Open the Cloud SDK Location preference page when the wizard closes
+REVIEW_DATASTORE_INDEXES_UPDATE_MESSAGE=The Development App Server has generated an update for the Datastore Indexes (datastore-indexes.xml). Would you like to review and merge the changes?
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/messages.properties]: "The local development server has updated the datastore indexes."

File: plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/LocalAppEngineServerLaunchConfigurationDelegate.java [hunk 2]
```diff
@@ -720,6 +722,26 @@ private CloudSdkDebugTarget(ILaunch launch, LocalAppEngineServerBehaviour server
       this.console = console;
     }
 
+    /** Check for an updated {@code datastore-index-auto.xml} in the {@code default} module. */
+    private void checkUpdatedDatastoreIndex(ILaunchConfiguration configuration) {
+      DatastoreIndexUpdateData update = DatastoreIndexUpdateData.detect(configuration, server);
+      if (update == null) {
+        return;
+      }
+      logger.info("datastore-indexes-auto.xml found " + update.datastoreIndexesAutoXml);
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/LocalAppEngineServerLaunchConfigurationDelegate.java]: do we really need this log message?

File: plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/DatastoreIndexesUpdatedStatusHandler.java [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+    Preconditions.checkArgument(source instanceof DatastoreIndexUpdateData);
+    DatastoreIndexUpdateData update = (DatastoreIndexUpdateData) source;
+    Preconditions.checkState(update.datastoreIndexesAutoXml != null);
+
+    if (DebugUITools.isPrivate(update.configuration)) {
+      return null;
+    }
+
+    IWorkbenchPage page = getActivePage();
+    // should consider using MessageDialogWithToggle? if original doesn't exist, should we copy into
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/DatastoreIndexesUpdatedStatusHandler.java]: todo

**Comment by briandealwis** [comment]
The `datastore-indexes-auto.xml` is described in https://cloud.google.com/appengine/docs/standard/java/config/indexconfig

File: plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/server/DatastoreIndexUpdateDataTest.java [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+import org.mockito.Mock;
+import org.mockito.runners.MockitoJUnitRunner;
+
+@RunWith(MockitoJUnitRunner.class)
+public class DatastoreIndexUpdateDataTest {
+  @Rule
+  public TemporaryFolder root = new TemporaryFolder();
+
+  @Mock
+  ILaunchConfiguration launchConfiguration;
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/server/DatastoreIndexUpdateDataTest.java]: These could all be private, I suspect

File: plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/messages.properties [hunk 5]
```diff
@@ -10,6 +10,11 @@ CREATE_APP_ENGINE_RUNTIME_WIZARD_DESCRIPTION=The App Engine Standard runtime req
 CREATE_APP_ENGINE_RUNTIME_WIZARD_TITLE=App Engine Standard Runtime
 SERVICE_CONFLICTS=Cannot add "{0}" as project "{1}" has App Engine Service ID: {2}
 OPEN_CLOUD_SDK_PREFERENCE_BUTTON=Open the Cloud SDK Location preference page when the wizard closes
+REVIEW_DATASTORE_INDEXES_UPDATE_MESSAGE=The local development server has updates for the \
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/messages.properties]: Per Google UI guidelines, "Avoid the subjunctive  'Do you want to save your changes?' is simpler to read than 'Would you like to save your changes?' The subjunctive rarely adds meaning, but is a symptom of elevated diction (using phrases, vocabulary, or syntax that is meant to impress as much as to communicate). Of course, even shorter and better than those examples is just, 'Save now?'"

So maybe "Review and merge the changes?"

**Pull request merged**